core/templates/services/playthrough.service.ts

```diff
@@ -17,383 +17,269 @@
-    }
+class CyclicStateTransitionsTracker {
+  pathOfUniqueStates: string[];
+  cycleOfStates: string[] = null;
+  numLoops: number = 0;
 
-    private recordStateTransitionInMultipleIncorrectIssueTracker(
-        destStateName: string): void {
-      if (PlaythroughService.multipleIncorrectStateName.num_times_incorrect <
-        ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD) {
-        PlaythroughService.multipleIncorrectStateName.state_name = (
-          destStateName);
-        PlaythroughService.multipleIncorrectStateName.num_times_incorrect = 0;
-      }
-    }
+  constructor(initStateName: string) {
+    this.pathOfUniqueStates = [initStateName];
+  }
 
-    private recordStateTransitionInCyclicIssueTracker(
-        destStateName: string): void {
-      if (PlaythroughService.cycleIdentifier.num_cycles <
-        ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD) {
-        if (PlaythroughService.visitedStates.indexOf(destStateName) !== -1) {
-          // Cycle identified.
-          let cycleStartIndex = PlaythroughService.visitedStates.indexOf(
-            destStateName);
-          PlaythroughService.visitedStates.push(destStateName);
-          let cycleString = PlaythroughService.visitedStates.slice(
-            cycleStartIndex, PlaythroughService.visitedStates.length
-          ).toString();
-          if (PlaythroughService.cycleIdentifier.cycle === cycleString) {
-            PlaythroughService.cycleIdentifier.num_cycles += 1;
-          } else {
-            PlaythroughService.cycleIdentifier.cycle = cycleString;
-            PlaythroughService.cycleIdentifier.num_cycles = 1;
-          }
-          PlaythroughService.visitedStates = [destStateName];
-        } else {
-          PlaythroughService.visitedStates.push(destStateName);
-        }
-      }
-    }
+  private makeCycle(collisionIndex: number): string[] {
+    const collision = this.pathOfUniqueStates[collisionIndex];
+    const cycleWithoutCollision = this.pathOfUniqueStates.slice(collisionIndex);
+    return [...cycleWithoutCollision, collision];
+  }
 
-    private isMultipleIncorrectSubmissionsIssue(): boolean {
-      return (
-        PlaythroughService.multipleIncorrectStateName.num_times_incorrect >=
-        ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD);
-    }
+  private currStateName(): string {
+    return this.pathOfUniqueStates[this.pathOfUniqueStates.length - 1];
+  }
 
-    private isCyclicStateTransitionsIssue(): boolean {
-      return PlaythroughService.cycleIdentifier.num_cycles >=
-        ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD;
-    }
+  foundAnIssue(): boolean {
+    return this.numLoops >= ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD;
+  }
 
-    private isEarlyQuitIssue(timeSpentInExpInSecs: number): boolean {
-      return timeSpentInExpInSecs <
-        ServicesConstants.EARLY_QUIT_THRESHOLD_IN_SECS;
+  recordStateTransition(destStateName: string): void {
```

Comment:
I'm actually a little confused about what's happening in this function. Could you perhaps add a longish docstring to explain?

In particular, one thing I'm wondering about is, if a state is visited twice, we don't seem to store that information. So if a user goes A --> B --> C --> A --> A --> A or A --> B --> C --> B --> C, what are those counted as? (I'm not sure what it _should_ be counted as.)

Probably worth adding some explanation here, therefore, to explain what exactly we are trying to detect.

Reply:
Done, let me know how it helps.

Reply:
It's great! I missed the last part about resetting the path and the docstring helped clarify that, thanks.

The only thing I'm wondering about is the B --> A --> A case. If you stay in the same state, is A --> A considered its own cycle (and therefore meriting a reset)? I feel it shouldn't merit a reset (cycles should include at least 2 different states), but if I'm reading the code right, the path does get reset to [A] at the end rather than staying as [B, A].

Reply:
But it shouldn't reset, right? It seems to me that you could miss a cycle in the case I described if, after the learner repeats A, they go to B.

More generally, you kind of want to treat B --> C --> A --> A --> B --> C --> A --> B as two instances of the same cycle.

Reply:
Done.